feat: add C ndarray interface and refactor implementation for `stats/base/dmeanpn`

gururaj1512 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

- adds C ndarray interface for `stats/base/dmeanpn`
- refactors JavaScript implementation
- refactors tests, docs, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add C ndarray interface and refactor implementation for `stats/base/dmeanpn`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +166,108 @@ console.log( v );
 
 <!-- /.examples -->
 
+<!-- C usage documentation. -->
```
incorrect readme structure, see #5138 for example

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
````diff
@@ -175,6 +166,108 @@ console.log( v );
+
+```c
+#include "stdlib/stats/base/dmeanpn.h"
+#include <stdint.h>
````
```suggestion
```
this header is not being used

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/benchmark/benchmark.js:
```diff
@@ -39,13 +48,7 @@ var dmeanpn = require( './../lib/dmeanpn.js' );
-	for ( i = 0; i < x.length; i++ ) {
-		x[ i ] = ( randu()*20.0 ) - 10.0;
-	}
+	var x = filledarrayBy( len, 'float64', rand );
```
just use `@stdlib/random/array/uniform` here no need to create another function

aayush0325 replied:
same comment for other benchmarks

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/examples/c/example.c:
remove the `stdint.h` header from here as it isn't being used

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/lib/dmeanpn.js:
```diff
@@ -40,35 +40,20 @@ var dapxsumpw = require( '@stdlib/blas/ext/base/dapxsumpw' );
-
-	return mu + c;
+function dmeanpn( N, x, strideX ) {
+	var ox = stride2offset( N, strideX );
+	return ndarray( N, x, strideX, ox );
```
```suggestion
	return ndarray( N, x, strideX, stride2offset( N, strideX ) );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/addon.c:
replace all instances of `stride` with `strideX` here

gururaj1512 committed:
chore: apply suggested changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
chore: update benchmark file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -51,36 +51,33 @@ The [arithmetic mean][arithmetic-mean] is defined as
-#### dmeanpn( N, x, stride )
+#### dmeanpn( N, x, strideX )
 
 Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-point strided array `x` using a two-pass error correction algorithm.
```
```suggestion
Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-point strided array using a two-pass error correction algorithm.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/README.md:
```diff
@@ -175,6 +166,123 @@ console.log( v );
+
+#### stdlib_strided_dmeanpn( N, \*X, strideX )
+
+Calculate the [arithmetic mean][arithmetic-mean] of a double-precision floating-point strided array using a two-pass error correction algorithm.
```
```suggestion
Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-point strided array using a two-pass error correction algorithm.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/docs/repl.txt:
```diff
@@ -19,8 +19,8 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/docs/repl.txt:
```diff
@@ -66,10 +63,10 @@
-    stride: integer
-        Index increment.
+    strideX: integer
+        Stride Length.
```
```suggestion
        Stride length.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/main.c:
```diff
@@ -33,30 +34,44 @@
 
 	// Compute an estimate for the mean:
-	mu = stdlib_strided_dsumpw( N, X, stride );
+	mu = stdlib_strided_dsumpw_ndarray( N, X, strideX, offsetX );
```
```suggestion
	mu = API_SUFFIX(stdlib_strided_dsumpw_ndarray)( N, X, strideX, offsetX );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmeanpn/src/main.c:
```diff
@@ -33,30 +34,44 @@
 
 	// Compute an error term:
-	c = stdlib_strided_dapxsumpw( N, -mu, X, stride );
+	c = stdlib_strided_dapxsumpw_ndarray( N, -mu, X, strideX, offsetX );
```
```suggestion
	c = API_SUFFIX(stdlib_strided_dapxsumpw_ndarray)( N, -mu, X, strideX, offsetX );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
